feat: articles on string types and object type inferences (#31)
LearningTypeScript/site: src/content/articles/for-in-string-types.mdx

---
date: "February 25 2023"
description: "Why TypeScript reports a `string` type instead of `keyof typeof` for some forms of loops over objects."
meta: keyof, loops, objects, strings
---

# Why `for...in` and `Object.keys()` Give `string` Types

> Why does TypeScript report a `string` type instead of `keyof typeof` the object when you use a `for...in` loop or `Object.keys()` on an object?
> 
> One of the most commonly asked questions in TypeScript help forums is why code like the following gives an error under the [`noImplicitAny` compiler option](https://www.typescriptlang.org/tsconfig#noImplicitAny).

Reviewer: [Nitpick/proofreading] Hmmm..., there seems to be some ambiguity in this part of your article. Maybe you can try the suggestion below to make it clearer.  
```suggestion
 One of the most commonly asked questions in TypeScript help forums is why code like the following gives an error under the [`noImplicitAny` compiler option](https://www.typescriptlang.org/tsconfig#noImplicitAny).
```

```ts twoslash
// @errors: 7053
const counts = {
  one: 1,
  two: 2,
};

for (const i in counts) {
  //       ^?
  console.log(counts[i]);
}
```

Shouldn't TypeScript know that `i` is type `"one" | "two"`, a.k.a. `keyof typeof counts`?
Why is it telling us that `i` is `string` and can't be used like `counts[i]`?

The answer is logical -though often inconvenient- and requires understanding a key characteristic of TypeScript's type system.

<!--truncate-->

## Objects May Have More Properties Than They Appear

TypeScript's type system is _structurally typed_: meaning any value that happens to satisfy a type is allowed to be used as a value of that type.
In other words, when you declare that a value is of a particular object type, you're telling TypeScript any object that just so happens to have those properties is allowed to be considered that type.

> As a result, when you write a `for...in` loop over the keys of an object, you often won't have a guarantee that the object's keys will only be from the type of the object.

Reviewer: [nitpick/proofreading] This paragraph could be a bit clearer. The first introductory phrase (As a result) creates a repetitive flow because the other phrase "when you write" conveys the same idea. 
```suggestion
As a result, writing a `for...in` loop over the keys of an object won't often guarantee that the object's keys will only be from the type of the object.
```

> Some other area of code might have provided an object with more properties that just so happens to match the structural type.

Reviewer: [Proofreading] Since this sentence is describing other parts of the code, consider turning "area" into a plural noun to convey this idea clearly. 
```suggestion
Some other areas of code might have provided an object with more properties that just so happens to match the structural type.
```

};

Object.keys(counts);
//     ^?
```

`Object.keys()` and `for...in` loops work almost the same way.
The main difference is that `Object.keys()` only looks at properties on the object itself, while `for...in` loops also look at properties on the object's `prototype` chain.

See [Workarounds](#workarounds) below for how you can get around this behavior in `Object.keys()` as well.

## Workarounds

Before you try to work around the design of the type checker, please pause a moment and reflect.
The situations mentioned earlier in this post are real problems in JavaScript and TypeScript code that have caused countless bugs in software.

**TypeScript is right to assume `string`s types.**

### `Object` APIs

TypeScript's type checking is designed to work with how JavaScript data typically flows and changes through an application.
Needing to wrestle with the type checker is often a byproduct of application code going against safe JavaScript practices -- such as the aforementioned issues with `for...in` loops receiving unexpected key strings.
In many cases, it might be good to listen to TypeScript and consider refactoring your code to use a JavaScript API better suited to the task.

> If all your code needs to do is get all the values in an object, an `Object.values()` API exists that returns those as an array.

Reviewer: [Proofreading/nitpick/question] I edited the sentence a bit to make it clearer. Is the below suggestion the message you want to convey? 
```suggestion
If all your code needs all an object's values, an `Object.values()` API exists to return those as an array.
```

You can `for...of` loop over them instead:

```ts twoslash
// @lib: dom,esnext

const counts = {
  one: 1,
  two: 2,
};

for (const [key, value] of Object.entries(counts)) {
  //        ^?
}
```

Alternately, if you do need keys, consider using the `Object.entries()` API instead.
It returns a `[string, T][]` where `T` is the type of values in an object:

```ts twoslash
// @lib: dom,esnext

const counts = {
  one: 1,
  two: 2,
};

for (const [key, value] of Object.entries(counts)) {
  //        ^?
}
```

Both `Object.entries()` and `Object.values()` allow you to write clean, idiomatic JavaScript without having to grapple with the type checker.

### `Map`

Taking an even larger step back, why is the application code looping over an object's keys and values?
Is the object meant to be used as a general key-value store?

> If so: consider using the built-in `Map` data structure instead.

Reviewer: [Proofreading/typo] It's always best to use a comma for introductory elements. 
```suggestion
If so, consider using the built-in `Map` data structure instead.
```

It's made specifically for key-value stores.

You can `for...of` over a `Map`'s values using `Array.from`:

```ts twoslash
// @lib: dom,esnext

const counts = new Map([
  ["one", 1],
  ["two", 2],
]);

for (const [key, value] of Array.from(counts)) {
  //        ^?
}
```

`Map`s have the added benefit over `{}` objects of allowing any object to be a key without stringifying them first (which objects do).

### Type Assertions

Still, there are times when a `for...in` loop or `Object.keys()` might likely be mostly safe to use for object keys.

> If you are 100% absolutely completely sure that your use case won't crash from some bizarre type mismatch now or in the future, you can always use an `as` assertion to cast the loop variable to the right type.

Reviewer: [Proofreading/nitpick] While mentioning "completely" and "absolutely" encourages devs to think deeply about the use case, I think saying "100% sure" conveys this point perfectly.  
```suggestion
If you are 100% sure that your use case won't crash from some bizarre type mismatch now or in the future, you can always use an `as` assertion to cast the loop variable to the right type.
```
